<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API documentation &mdash; PascalX 0.0.4 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> PascalX
            <img src="_static/Xlogo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#genome">Genome</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.genome.genome"><code class="docutils literal notranslate"><span class="pre">genome</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.genome.genome.gene_info"><code class="docutils literal notranslate"><span class="pre">genome.gene_info()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.genome.genome.get_ensembl_annotation"><code class="docutils literal notranslate"><span class="pre">genome.get_ensembl_annotation()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.genome.genome.load_genome"><code class="docutils literal notranslate"><span class="pre">genome.load_genome()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#genescorer">Genescorer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pathwayscorer">Pathwayscorer</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.pathway.chi2rank"><code class="docutils literal notranslate"><span class="pre">chi2rank</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2rank.score"><code class="docutils literal notranslate"><span class="pre">chi2rank.score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2rank.get_sigpathways"><code class="docutils literal notranslate"><span class="pre">chi2rank.get_sigpathways()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2rank.load_modules"><code class="docutils literal notranslate"><span class="pre">chi2rank.load_modules()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2rank.set_genescorer"><code class="docutils literal notranslate"><span class="pre">chi2rank.set_genescorer()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.pathway.chi2perm"><code class="docutils literal notranslate"><span class="pre">chi2perm</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2perm.score"><code class="docutils literal notranslate"><span class="pre">chi2perm.score()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2perm.get_sigpathways"><code class="docutils literal notranslate"><span class="pre">chi2perm.get_sigpathways()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2perm.load_modules"><code class="docutils literal notranslate"><span class="pre">chi2perm.load_modules()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.pathway.chi2perm.set_genescorer"><code class="docutils literal notranslate"><span class="pre">chi2perm.set_genescorer()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#xscorer">Xscorer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genexprscorer">Genexprscorer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#snp-database">SNP database</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.snpdb.db"><code class="docutils literal notranslate"><span class="pre">db</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.open"><code class="docutils literal notranslate"><span class="pre">db.open()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.insert"><code class="docutils literal notranslate"><span class="pre">db.insert()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.get"><code class="docutils literal notranslate"><span class="pre">db.get()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getSNPatPos"><code class="docutils literal notranslate"><span class="pre">db.getSNPatPos()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getPosatSNPs"><code class="docutils literal notranslate"><span class="pre">db.getPosatSNPs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getSNPs"><code class="docutils literal notranslate"><span class="pre">db.getSNPs()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getSNPKeys"><code class="docutils literal notranslate"><span class="pre">db.getSNPKeys()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getKeys"><code class="docutils literal notranslate"><span class="pre">db.getKeys()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.getSortedKeys"><code class="docutils literal notranslate"><span class="pre">db.getSortedKeys()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.snpdb.db.close"><code class="docutils literal notranslate"><span class="pre">db.close()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#reference-panel">Reference panel</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.refpanel.refpanel"><code class="docutils literal notranslate"><span class="pre">refpanel</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.refpanel.refpanel.load_pos_reference"><code class="docutils literal notranslate"><span class="pre">refpanel.load_pos_reference()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.refpanel.refpanel.load_snp_reference"><code class="docutils literal notranslate"><span class="pre">refpanel.load_snp_reference()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.refpanel.refpanel.set_refpanel"><code class="docutils literal notranslate"><span class="pre">refpanel.set_refpanel()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.refpanel.refpanel.getSNPtoChrMap"><code class="docutils literal notranslate"><span class="pre">refpanel.getSNPtoChrMap()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.refpanel.refpanel.getChrSNPs"><code class="docutils literal notranslate"><span class="pre">refpanel.getChrSNPs()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#mapper">Mapper</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#PascalX.mapper.mapper"><code class="docutils literal notranslate"><span class="pre">mapper</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#PascalX.mapper.mapper.load_mapping"><code class="docutils literal notranslate"><span class="pre">mapper.load_mapping()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">PascalX</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">API documentation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/PascalX.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-documentation">
<h1>API documentation<a class="headerlink" href="#api-documentation" title="Permalink to this heading"></a></h1>
<p>The PascalX package comes with the following user modules:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#genome">Genome</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.genome</span></code>)</p></li>
<li><p><a class="reference internal" href="#genescorer">Genescorer</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.genescorer</span></code>)</p></li>
<li><p><a class="reference internal" href="#pathwayscorer">Pathwayscorer</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.pathway</span></code>)</p></li>
<li><p><a class="reference internal" href="#xscorer">Xscorer</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.xscorer</span></code>)</p></li>
<li><p><a class="reference internal" href="#genexprscorer">Genexprscorer</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.genexpr</span></code>)</p></li>
</ul>
<p>Internal modules:</p>
<ul class="simple">
<li><p><a class="reference internal" href="#snpdb">SNPdb</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.snpdb</span></code>)</p></li>
<li><p><a class="reference internal" href="#refpanel">RefPanel</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.refpanel</span></code>)</p></li>
<li><p><a class="reference internal" href="#mapper">Mapper</a> (<code class="code docutils literal notranslate"><span class="pre">PascalX.mapper</span></code>)</p></li>
</ul>
<hr class="docutils" />
<section id="genome">
<span id="id1"></span><h2>Genome<a class="headerlink" href="#genome" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PascalX.genome.genome">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.genome.</span></span><span class="sig-name descname"><span class="pre">genome</span></span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/genome.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.genome.genome" title="Permalink to this definition"></a></dt>
<dd><p>This class handles the genome annotation. It provides functionality for import of data from text files and automatic download of annotation data from ensembl.org.</p>
<dl class="py method">
<dt class="sig sig-object py" id="PascalX.genome.genome.gene_info">
<span class="sig-name descname"><span class="pre">gene_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/genome.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.genome.genome.gene_info" title="Permalink to this definition"></a></dt>
<dd><p>Prints the loaded information for given gene</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>gene</strong> (<em>string</em>) – Symbol of gene to query</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.genome.genome.get_ensembl_annotation">
<span class="sig-name descname"><span class="pre">get_ensembl_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'protein_coding'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GRCh38'</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/genome.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.genome.genome.get_ensembl_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Gene annotation download function for ensembl.org BioMart data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>string</em>) – File to store downloaded annotation in</p></li>
<li><p><strong>genetype</strong> (<em>string</em>) – Comma separated list of BioMart genetypes to download (protein_coding, pseudogene, …)</p></li>
<li><p><strong>version</strong> (<em>string</em>) – GRCh37 | CRCh38</p></li>
</ul>
</dd>
</dl>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">PascalX.genome</span> <span class="kn">import</span> <span class="n">genome</span>
<span class="n">G</span> <span class="o">=</span> <span class="n">genome</span><span class="p">()</span>
<span class="n">G</span><span class="o">.</span><span class="n">get_ensemble_annotation</span><span class="p">(</span><span class="s1">&#39;ensemble_hg38.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.genome.genome.load_genome">
<span class="sig-name descname"><span class="pre">load_genome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ccol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csymb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cstx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cetx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrStart</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splitchr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">NAgeneid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'n/a'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">useNAgenes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/genome.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.genome.genome.load_genome" title="Permalink to this definition"></a></dt>
<dd><p>Imports gene annotation from text file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>text</em>) – File to load</p></li>
<li><p><strong>ccol</strong> (<em>int</em>) – Column containing chromosome number</p></li>
<li><p><strong>cid</strong> (<em>int</em>) – Column containing gene id</p></li>
<li><p><strong>csymb</strong> (<em>int</em>) – Column containing gene symbol</p></li>
<li><p><strong>cstx</strong> (<em>int</em>) – Column containing transcription start</p></li>
<li><p><strong>cetx</strong> (<em>int</em>) – Column containing transcription end</p></li>
<li><p><strong>cs</strong> (<em>int</em>) – Column containing strand</p></li>
<li><p><strong>cb</strong> (<em>int</em>) – Column containing band (None if not supplied)</p></li>
<li><p><strong>chrStart</strong> (<em>int</em>) – Number of leading characters to skip in ccol</p></li>
<li><p><strong>splitchr</strong> (<em>string</em>) – Character used to separate columns in text file</p></li>
<li><p><strong>NAgeneid</strong> (<em>string</em>) – Identifier for not available gene id</p></li>
<li><p><strong>useNAgenes</strong> (<em>bool</em>) – Import genes without gene id</p></li>
<li><p><strong>header</strong> (<em>bool</em>) – First line is header</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt><strong>Internal:</strong></dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">_GENEID</span></code> (dict) - Contains the raw data with gene id (cid) as key</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_GENESYMB</span></code> (dict) - Mapping from gene symbols (csymb) to gene ids (cid)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_GENEIDtoSYMB</span></code> (dict) - Mapping from gene ids (cid) to gene symbols (csymb)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_CHR</span></code> (dict) - Mapping from chromosomes to list of gene symbols</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_BAND</span></code> (dict) - Mapping from band to list of gene symbols</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">_SKIPPED</span></code> (dict) - Genes (cid) which could not be imported</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An unique gene id is automatically generated for n/a gene ids if <code class="docutils literal notranslate"><span class="pre">useNAgenes=true</span></code>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Identical gene ids occuring in more than one row are merged to a single gene, if on same chromosome and positional gap is &lt; 1Mb. The strand is taken from the longer segment.</p>
</div>
</dd></dl>

</dd></dl>

<hr class="docutils" />
</section>
<section id="genescorer">
<span id="id2"></span><h2>Genescorer<a class="headerlink" href="#genescorer" title="Permalink to this heading"></a></h2>
</section>
<section id="pathwayscorer">
<span id="id3"></span><h2>Pathwayscorer<a class="headerlink" href="#pathwayscorer" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PascalX.pathway.chi2rank">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.pathway.</span></span><span class="sig-name descname"><span class="pre">chi2rank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genescorer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mergedist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fuse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2rank" title="Permalink to this definition"></a></dt>
<dd><p>Pathway scoring via chi2 of ranked gene p-values. Nearby genes can be merged to form meta-genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genescorer</strong> (<em>genescorer</em>) – The initialized genescorer to use to re-compute fused genes</p></li>
<li><p><strong>mergedist</strong> (<em>int</em>) – Maximum inbetween distance of genes to merge</p></li>
<li><p><strong>fuse</strong> (<em>bool</em>) – Fuse nearby genes to meta-genes</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2rank.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modules</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'saddle'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reqacc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nobar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrs_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">autorescore</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2rank.score" title="Permalink to this definition"></a></dt>
<dd><p>Scores a set of pathways/modules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modules</strong> (<em>list</em>) – List of modules to score</p></li>
<li><p><strong>samples</strong> (<em>int</em>) – # of random gene sets to draw</p></li>
<li><p><strong>method</strong> (<em>string</em>) – Method to use to evaluate tail probability (‘auto’,’davies’,’ruben’,’satterthwaite’,’pearson’,’saddle’)</p></li>
<li><p><strong>mode</strong> (<em>string</em>) – Precision mode to use (‘’,’128b’,’100d’,’auto’)</p></li>
<li><p><strong>reqacc</strong> (<em>float</em>) – requested accuracy</p></li>
<li><p><strong>autorescore</strong> (<em>bool</em>) – Automatically try to re-score failed genes</p></li>
<li><p><strong>nobar</strong> (<em>bool</em>) – Show progress bar</p></li>
<li><p><strong>genes_only</strong> (<em>bool</em>) – Compute only (fused)-genescores (accessible via genescorer method)</p></li>
<li><p><strong>chrs_only</strong> (<em>list</em>) – Only consider genes on listed chromosomes. None for all</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2rank.get_sigpathways">
<span class="sig-name descname"><span class="pre">get_sigpathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RESULT</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2rank.get_sigpathways" title="Permalink to this definition"></a></dt>
<dd><p>Prints significant pathways in the result set</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>RESULT</strong> (<em>list</em>) – Return of a pathwayscorer</p></li>
<li><p><strong>cutoff</strong> (<em>float</em>) – Significance threshold to print pathways</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2rank.load_modules">
<span class="sig-name descname"><span class="pre">load_modules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2rank.load_modules" title="Permalink to this definition"></a></dt>
<dd><p>Load modules from tab separated file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>string</em>) – path/filename</p></li>
<li><p><strong>ncol</strong> (<em>int</em>) – Column with name of module</p></li>
<li><p><strong>fcol</strong> (<em>int</em>) – Column with first gene (symbol) in module. Remaining genes have to follow tab (  ) separated</p></li>
<li><p><strong>symbol</strong> (<em>bool</em>) – Genes are given as gene symbols (False requires genome to be set in genescorer)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2rank.set_genescorer">
<span class="sig-name descname"><span class="pre">set_genescorer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">S</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2rank.set_genescorer" title="Permalink to this definition"></a></dt>
<dd><p>Set the genescorer to use for re-computing p-values of fused genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>S</strong> (<em>genescorer</em>) – The initialized genescorer to use</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="PascalX.pathway.chi2perm">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.pathway.</span></span><span class="sig-name descname"><span class="pre">chi2perm</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genescorer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mergedist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fuse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2perm" title="Permalink to this definition"></a></dt>
<dd><p>Pathway scoring via testing summed inverse chi2 transformed gene p-values against equally size random samples of gene sets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genescorer</strong> (<em>genescorer</em>) – The initialized genescorer to use to re-compute fused genes</p></li>
<li><p><strong>mergedist</strong> (<em>int</em>) – Maximum inbetween distance of genes to merge</p></li>
<li><p><strong>fuse</strong> (<em>bool</em>) – Fuse nearby genes to meta-genes</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Genes in the background gene sets are NOT fused.</p>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2perm.score">
<span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modules</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'saddle'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reqacc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nobar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">autorescore</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2perm.score" title="Permalink to this definition"></a></dt>
<dd><p>Scores a set of pathways/modules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>modules</strong> (<em>list</em>) – List of modules to score</p></li>
<li><p><strong>samples</strong> (<em>int</em>) – # of random gene sets to draw</p></li>
<li><p><strong>method</strong> (<em>string</em>) – Method to use to evaluate tail probability (‘auto’,’davies’,’ruben’,’satterthwaite’,’pearson’,’saddle’)</p></li>
<li><p><strong>mode</strong> (<em>string</em>) – Precision mode to use (‘’,’128b’,’100d’,’auto’)</p></li>
<li><p><strong>reqacc</strong> (<em>float</em>) – requested accuracy</p></li>
<li><p><strong>nobar</strong> (<em>bool</em>) – Show progress bar</p></li>
<li><p><strong>autorescore</strong> (<em>bool</em>) – Automatically try to re-score failed genes</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2perm.get_sigpathways">
<span class="sig-name descname"><span class="pre">get_sigpathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">RESULT</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2perm.get_sigpathways" title="Permalink to this definition"></a></dt>
<dd><p>Prints significant pathways in the result set</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>RESULT</strong> (<em>list</em>) – Return of a pathwayscorer</p></li>
<li><p><strong>cutoff</strong> (<em>float</em>) – Significance threshold to print pathways</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2perm.load_modules">
<span class="sig-name descname"><span class="pre">load_modules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2perm.load_modules" title="Permalink to this definition"></a></dt>
<dd><p>Load modules from tab separated file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>string</em>) – path/filename</p></li>
<li><p><strong>ncol</strong> (<em>int</em>) – Column with name of module</p></li>
<li><p><strong>fcol</strong> (<em>int</em>) – Column with first gene (symbol) in module. Remaining genes have to follow tab (  ) separated</p></li>
<li><p><strong>symbol</strong> (<em>bool</em>) – Genes are given as gene symbols (False requires genome to be set in genescorer)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.pathway.chi2perm.set_genescorer">
<span class="sig-name descname"><span class="pre">set_genescorer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">S</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/pathway.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.pathway.chi2perm.set_genescorer" title="Permalink to this definition"></a></dt>
<dd><p>Set the genescorer to use for re-computing p-values of fused genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>S</strong> (<em>genescorer</em>) – The initialized genescorer to use</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="xscorer">
<span id="id4"></span><h2>Xscorer<a class="headerlink" href="#xscorer" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
</section>
<section id="genexprscorer">
<span id="id5"></span><h2>Genexprscorer<a class="headerlink" href="#genexprscorer" title="Permalink to this heading"></a></h2>
<hr class="docutils" />
</section>
<section id="snp-database">
<span id="snpdb"></span><h2>SNP database<a class="headerlink" href="#snp-database" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PascalX.snpdb.db">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.snpdb.</span></span><span class="sig-name descname"><span class="pre">db</span></span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db" title="Permalink to this definition"></a></dt>
<dd><p>Class for handling storage of the raw genotype data. The data is indexed and stored for each chromosome individually as zlib compressed pickle. The indexing allows fast random access via SNP ids or positions.</p>
<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.open">
<span class="sig-name descname"><span class="pre">open</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.open" title="Permalink to this definition"></a></dt>
<dd><p>Opens storage file. A new file is created if not exists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>filename</strong> (<em>string</em>) – Name to use for the storage file</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.insert" title="Permalink to this definition"></a></dt>
<dd><p>Stores set of rows into the file storage</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>dict</em>) – First storage key is the key in the dictionary. Second storage key is the first element of the inner list.</p>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If all insert calls are done, the close function has to be called once to make the index persistent.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.get">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.get" title="Permalink to this definition"></a></dt>
<dd><p>Returns all stored data for a set of SNPs indexed via positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pos</strong> (<em>list</em>) – Positions of SNPs to retrieve</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getSNPatPos">
<span class="sig-name descname"><span class="pre">getSNPatPos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getSNPatPos" title="Permalink to this definition"></a></dt>
<dd><p>Returns SNP id at position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pos</strong> (<em>list</em>) – Positions of SNPs to retrieve</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getPosatSNPs">
<span class="sig-name descname"><span class="pre">getPosatSNPs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">snpids</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getPosatSNPs" title="Permalink to this definition"></a></dt>
<dd><p>Returns the position corresponding to a snpid
WARNING: Inefficient</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getSNPs">
<span class="sig-name descname"><span class="pre">getSNPs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">snps</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getSNPs" title="Permalink to this definition"></a></dt>
<dd><p>Returns all stored data for a set of SNPs indexed via SNP ids</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>snp</strong> (<em>list</em>) – ids of SNPs to retrieve</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getSNPKeys">
<span class="sig-name descname"><span class="pre">getSNPKeys</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getSNPKeys" title="Permalink to this definition"></a></dt>
<dd><p>Returns the SNP ids in storage</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getKeys">
<span class="sig-name descname"><span class="pre">getKeys</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getKeys" title="Permalink to this definition"></a></dt>
<dd><p>Returns SNP positions in storage</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.getSortedKeys">
<span class="sig-name descname"><span class="pre">getSortedKeys</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.getSortedKeys" title="Permalink to this definition"></a></dt>
<dd><p>Returns a sorted list of SNP positions in storage</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.snpdb.db.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/snpdb.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.snpdb.db.close" title="Permalink to this definition"></a></dt>
<dd><p>Closes open storage file.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>After all inserts are done this function has to be called once to re-generate the index and close the storage file.</p>
</div>
</dd></dl>

</dd></dl>

<hr class="docutils" />
</section>
<section id="reference-panel">
<span id="refpanel"></span><h2>Reference panel<a class="headerlink" href="#reference-panel" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.refpanel.</span></span><span class="sig-name descname"><span class="pre">refpanel</span></span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel.load_pos_reference">
<span class="sig-name descname"><span class="pre">load_pos_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel.load_pos_reference" title="Permalink to this definition"></a></dt>
<dd><p>Returns a snpdb object for a chromosome and a sorted list of SNP positions on the chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cr</strong> (<em>int</em>) – Chromosome number</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel.load_snp_reference">
<span class="sig-name descname"><span class="pre">load_snp_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel.load_snp_reference" title="Permalink to this definition"></a></dt>
<dd><p>Returns a snpdb object for a chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cr</strong> (<em>int</em>) – Chromosome number</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel.set_refpanel">
<span class="sig-name descname"><span class="pre">set_refpanel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keepfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualityT</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">SNPonly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chrlist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sourcefilename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regEx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nobar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel.set_refpanel" title="Permalink to this definition"></a></dt>
<dd><p>Sets the reference panel to use</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>string</em>) – /path/filename (without .chr#.db ending)</p></li>
<li><p><strong>parallel</strong> (<em>int</em>) – Number of cores to use for parallel import of reference panel</p></li>
<li><p><strong>keepfile</strong> (<em>string</em>) – [only for .vcf] File with sample ids (one per line) to keep.  None to keep all.</p></li>
<li><p><strong>qualityT</strong> (<em>int</em>) – [only for .vcf] Quality threshold for variant to keep (None to ignore)</p></li>
<li><p><strong>SNPonly</strong> (<em>bool</em>) – [only for .vcf] Load only SNPs</p></li>
<li><p><strong>chrlist</strong> (<em>list</em>) – List of chromosomes to import. (None to import 1-22)</p></li>
<li><p><strong>sourcefilename</strong> (<em>string</em>) – /path/filename (without .chr#. ending) of .tped | .vcf files. None to use same as filename</p></li>
<li><p><strong>regEx</strong> (<em>string</em>) – Regular expression to filter sample ids. First capture group is kept. [only for .vcf]</p></li>
<li><p><strong>nobar</strong> (<em>bool</em>) – Show progress bar (updates only if a chromosome finished)</p></li>
</ul>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>One file per chromosome with ending .chr#.db required (#: 1-22). If imported reference panel is not present, PascalX will automatically try to import from .chr#.tped.gz or .chr#.vcf.gz files.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Alleles (under .vcf import) are stored internally in the order [ALT,REF].</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel.getSNPtoChrMap">
<span class="sig-name descname"><span class="pre">getSNPtoChrMap</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel.getSNPtoChrMap" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary mapping SNP id to corresponding chromosome number</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="PascalX.refpanel.refpanel.getChrSNPs">
<span class="sig-name descname"><span class="pre">getChrSNPs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cr</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/refpanel.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.refpanel.refpanel.getChrSNPs" title="Permalink to this definition"></a></dt>
<dd><p>Returns SNP ids for a chromosome</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cr</strong> (<em>int</em>) – Chromosome number</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<hr class="docutils" />
</section>
<section id="mapper">
<span id="id6"></span><h2>Mapper<a class="headerlink" href="#mapper" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="PascalX.mapper.mapper">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">PascalX.mapper.</span></span><span class="sig-name descname"><span class="pre">mapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/mapper.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.mapper.mapper" title="Permalink to this definition"></a></dt>
<dd><dl class="py method">
<dt class="sig sig-object py" id="PascalX.mapper.mapper.load_mapping">
<span class="sig-name descname"><span class="pre">load_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a1col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a2col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pcol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'\t'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pfilter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference external" href="https://github.com/BergmannLab/PascalX/tree/main/python/PascalX/mapper.py"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#PascalX.mapper.mapper.load_mapping" title="Permalink to this definition"></a></dt>
<dd><p>Loads a SNP to gene mapping</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file</strong> (<em>string</em>) – File to load</p></li>
<li><p><strong>gcol</strong> (<em>int</em>) – Column with gene id</p></li>
<li><p><strong>rcol</strong> (<em>int</em>) – Column with SNP id</p></li>
<li><p><strong>wcol</strong> (<em>int</em>) – Column with weight (None for none)</p></li>
<li><p><strong>a1col</strong> (<em>int</em>) – Column of alternate allele (None for ignoring alleles)</p></li>
<li><p><strong>a2col</strong> (<em>int</em>) – Column of reference allele (None for ignoring alleles)</p></li>
<li><p><strong>bcol</strong> (<em>int</em>) – Column with additional weight (None for none)</p></li>
<li><p><strong>pcol</strong> (<em>int</em>) – Column with pvalue (For None p-value is taken from .load_GWAS data)</p></li>
<li><p><strong>delimiter</strong> (<em>string</em>) – Character used to separate columns</p></li>
<li><p><strong>header</strong> (<em>bool</em>) – Header present</p></li>
<li><p><strong>pfilter</strong> (<em>float</em>) – Only include rows with pcol &lt; pfilter</p></li>
<li><p><strong>symbol</strong> (<em>bool</em>) – Gene id are gene symbols (requires genome to be set on init)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021-2022, Bergmann lab and contributors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>