Search.setIndex({"docnames": ["PascalX", "index", "install", "usage"], "filenames": ["PascalX.rst", "index.rst", "install.rst", "usage.rst"], "titles": ["API documentation", "Welcome to PascalX\u2019s documentation!", "Installation", "Usage"], "terms": {"The": [0, 1, 2, 3], "pascalx": [0, 2, 3], "packag": [0, 1], "come": 0, "follow": [0, 1, 2, 3], "user": 0, "modul": [0, 3], "pathwai": [0, 1], "genexpr": [0, 1, 3], "intern": 0, "snpdb": 0, "refpanel": [0, 1, 3], "class": [0, 3], "sourc": [0, 1], "thi": [0, 1, 3], "handl": 0, "annot": [0, 3], "It": [0, 3], "provid": 0, "function": [0, 1], "import": [0, 3], "data": [0, 1, 2, 3], "from": [0, 3], "text": [0, 3], "file": [0, 3], "automat": [0, 3], "download": [0, 3], "ensembl": [0, 3], "org": 0, "gene_info": 0, "gene": [0, 1], "print": 0, "load": [0, 3], "inform": [0, 3], "given": [0, 1, 3], "paramet": [0, 3], "string": [0, 3], "symbol": [0, 3], "queri": 0, "get_ensembl_annot": [0, 3], "filenam": [0, 3], "genetyp": [0, 3], "protein_cod": [0, 3], "version": [0, 3], "grch38": [0, 3], "biomart": [0, 3], "store": [0, 3], "comma": [0, 3], "separ": [0, 3], "list": [0, 3], "pseudogen": 0, "grch37": [0, 3], "crch38": 0, "exampl": [0, 3], "g": [0, 2, 3], "get_ensemble_annot": 0, "ensemble_hg38": 0, "txt": [0, 3], "load_genom": [0, 3], "ccol": 0, "1": [0, 3], "cid": 0, "0": [0, 2, 3], "csymb": 0, "5": [0, 3], "cstx": 0, "2": [0, 1, 3], "cetx": 0, "3": [0, 3], "c": [0, 1], "4": [0, 3], "cb": 0, "none": [0, 3], "chrstart": 0, "splitchr": 0, "t": [0, 2], "nageneid": 0, "n": [0, 3], "usenagen": 0, "fals": [0, 3], "header": [0, 2, 3], "int": 0, "column": [0, 3], "contain": [0, 3], "chromosom": [0, 1, 3], "number": [0, 2, 3], "id": [0, 3], "transcript": [0, 3], "start": [0, 2, 3], "end": [0, 3], "strand": 0, "band": 0, "suppli": [0, 3], "lead": 0, "charact": 0, "skip": [0, 3], "us": [0, 1, 2, 3], "identifi": 0, "avail": [0, 2, 3], "bool": 0, "without": [0, 3], "first": [0, 3], "line": [0, 1], "i": [0, 1, 2, 3], "_geneid": 0, "dict": 0, "raw": [0, 3], "kei": 0, "_genesymb": 0, "map": [0, 3], "_geneidtosymb": 0, "_chr": 0, "_band": 0, "_skip": 0, "which": [0, 1, 3], "could": 0, "an": [0, 1, 3], "uniqu": 0, "gener": [0, 1, 3], "true": [0, 3], "ident": 0, "occur": 0, "more": [0, 3], "than": [0, 3], "one": [0, 3], "row": [0, 3], "ar": [0, 3], "merg": 0, "singl": 0, "same": [0, 3], "posit": [0, 3], "gap": 0, "1mb": 0, "taken": 0, "longer": 0, "segment": 0, "chi2rank": [0, 1, 3], "mergedist": 0, "100000": 0, "fuse": [0, 3], "score": [0, 1], "via": [0, 1, 2, 3], "chi2": 0, "rank": [0, 3], "p": [0, 1, 2, 3], "valu": [0, 1, 3], "nearbi": 0, "can": [0, 1, 2, 3], "form": 0, "meta": [0, 3], "initi": [0, 3], "re": [0, 3], "comput": [0, 1, 2, 3], "maximum": 0, "inbetween": 0, "distanc": [0, 3], "method": [0, 3], "saddl": [0, 3], "mode": [0, 2], "auto": [0, 3], "reqacc": 0, "1e": 0, "100": [0, 1], "parallel": [0, 1, 2, 3], "nobar": 0, "genes_onli": 0, "chrs_onli": 0, "set": [0, 2, 3], "sampl": [0, 3], "random": [0, 1, 3], "draw": 0, "evalu": [0, 3], "tail": 0, "probabl": 0, "davi": [0, 1], "ruben": [0, 1, 3], "satterthwait": 0, "pearson": [0, 3], "precis": [0, 1, 2, 3], "128b": 0, "100d": 0, "float": 0, "request": 0, "accuraci": 0, "show": 0, "progress": 0, "bar": 0, "onli": [0, 3], "access": [0, 1, 3], "consid": 0, "all": [0, 1, 2, 3], "get_sigpathwai": 0, "result": [0, 1, 3], "cutoff": [0, 3], "0001": 0, "signific": [0, 3], "return": [0, 3], "threshold": 0, "load_modul": [0, 3], "ncol": [0, 3], "fcol": [0, 3], "tab": [0, 3], "path": [0, 2, 3], "name": [0, 3], "remain": 0, "have": [0, 3], "requir": [0, 2, 3], "set_genescor": 0, "": [0, 3], "chi2perm": [0, 1, 3], "test": [0, 1, 3], "sum": [0, 3], "invers": [0, 3], "transform": [0, 3], "against": [0, 3], "equal": [0, 3], "size": [0, 2, 3], "background": 0, "NOT": 0, "db": [0, 1], "storag": 0, "genotyp": [0, 3], "index": 0, "each": [0, 3], "individu": [0, 3], "zlib": 0, "compress": [0, 3], "pickl": [0, 3], "allow": [0, 1, 2], "fast": [0, 1], "open": [0, 1, 2], "A": [0, 2, 3], "new": [0, 3], "creat": 0, "exist": [0, 3], "insert": 0, "dictionari": [0, 3], "second": [0, 3], "element": 0, "inner": 0, "If": [0, 1, 3], "call": [0, 3], "done": [0, 3], "close": [0, 3], "ha": [0, 3], "onc": [0, 3], "make": [0, 1, 2], "persist": [0, 2, 3], "get": 0, "po": 0, "retriev": 0, "getsnpatpo": 0, "getposatsnp": 0, "snpid": 0, "correspond": [0, 3], "warn": 0, "ineffici": 0, "getsnp": 0, "getsnpkei": 0, "getkei": 0, "getsortedkei": 0, "sort": 0, "after": [0, 2, 3], "load_pos_refer": 0, "cr": 0, "keep_idx": 0, "object": [0, 3], "load_snp_refer": 0, "set_refpanel": 0, "keepfil": [0, 3], "qualityt": 0, "snponli": 0, "chrlist": 0, "sourcefilenam": 0, "regex": 0, "chr": [0, 3], "core": [0, 3], "vcf": [0, 3], "per": [0, 3], "keep": [0, 3], "qualiti": 0, "variant": [0, 3], "ignor": 0, "22": [0, 3], "tped": [0, 3], "regular": 0, "express": [0, 3], "filter": [0, 3], "captur": 0, "group": 0, "kept": [0, 3], "updat": 0, "finish": 0, "One": 0, "present": 0, "try": [0, 3], "gz": [0, 3], "allel": [0, 3], "under": [0, 1, 3], "order": [0, 3], "alt": 0, "ref": 0, "getsnptochrmap": 0, "getchrsnp": 0, "load_map": [0, 3], "gcol": [0, 3], "rcol": [0, 3], "wcol": 0, "a1col": [0, 3], "a2col": [0, 3], "bcol": [0, 3], "pcol": [0, 3], "delimit": [0, 3], "pfilter": 0, "weight": [0, 1], "altern": [0, 3], "addit": [0, 1, 3], "pvalu": 0, "For": [0, 1, 2, 3], "load_gwa": [0, 3], "includ": [0, 3], "init": 0, "python3": [1, 2, 3], "librari": [1, 2, 3], "high": [1, 3], "gwa": [0, 1, 2, 3], "summari": [0, 1, 3], "statist": [0, 1, 3], "aggreg": [1, 3], "snp": 1, "pascal": [1, 3], "methodologi": [1, 3], "base": [0, 1, 3], "chi": [1, 3], "cummul": 1, "distribut": [1, 3], "calcul": [0, 1, 3], "approxim": [1, 3], "exactli": 1, "multi": [1, 3], "implement": [0, 1], "algorithm": [0, 1, 3], "appli": 1, "modern": 1, "uk": 1, "biobank": 1, "scale": 1, "In": [1, 3], "offer": [1, 3], "novel": 1, "coher": [0, 1, 3], "between": [0, 1, 3], "two": [0, 1, 3], "differ": [0, 1, 3], "level": 1, "product": 1, "normal": [0, 1, 3], "describ": [1, 3], "here": [1, 3], "highlight": 1, "cdf": [1, 3], "up": 1, "digit": 1, "over": [0, 1], "gpu": [0, 1], "support": [1, 2, 3], "acceler": 1, "linear": 1, "algebra": 1, "oper": [1, 3], "refer": [1, 2, 3], "panel": [1, 2, 3], "genom": [1, 3], "own": [1, 3], "databas": 1, "wise": [1, 3], "select": [1, 3], "extern": [1, 3], "tissu": [0, 1], "enrich": [0, 1, 3], "experiment": [1, 3], "research": 1, "tool": 1, "No": 1, "warranti": 1, "guarante": 1, "whatsoev": 1, "its": 1, "correct": [0, 1], "you": [0, 1, 2, 3], "should": [1, 2, 3], "perform": [0, 1, 3], "your": [1, 2, 3], "consist": 1, "check": 1, "impli": 1, "instal": 1, "platform": [1, 3], "docker": [1, 3], "debian": 1, "ubuntu": 1, "mac": 1, "window": [0, 1, 3], "usag": [0, 1], "command": [1, 2], "interfac": 1, "x": 1, "api": [1, 3], "genescor": [1, 3], "pathwayscor": 1, "xscorer": [1, 3], "genexprscor": 1, "mapper": 1, "agplv3": 1, "pleas": [1, 2, 3], "cite": 1, "doi": [1, 3], "10": [0, 1, 3], "5281": 1, "zenodo": 1, "4429922": 1, "scorer": [0, 1, 3], "work": [1, 2], "krefl": 1, "d": [1, 2], "bergmann": 1, "cross": [0, 1, 3], "plo": 1, "biologi": 1, "18": [0, 1], "9": [0, 1], "e1010517": 1, "1371": 1, "journal": 1, "pcbi": 1, "1010517": 1, "technic": 1, "issu": 1, "bug": 1, "report": 1, "etc": [1, 2], "ticket": 1, "github": [1, 2], "page": 1, "inquiri": 1, "main": 1, "contributor": 1, "daniel": 1, "unil": 1, "ch": 1, "build": 2, "imag": 2, "git": 2, "clone": 2, "http": 2, "com": 2, "bergmannlab": 2, "cd": 2, "latest": 2, "run": [0, 2, 3], "interact": [2, 3], "host": 2, "directori": 2, "workdir": 2, "mount": 2, "src": 2, "target": 2, "type": [0, 2, 3], "bind": 2, "8888": 2, "bash": [2, 3], "stop": 2, "exlus": 2, "jupyt": 2, "notebook": 2, "ip": 2, "root": 2, "machin": 2, "browser": 2, "visit": 2, "localhost": 2, "alloc": 2, "suffici": [0, 2], "system": 2, "resourc": 2, "runtim": [2, 3], "minimum": 2, "8gb": 2, "ram": 2, "made": 2, "full": [0, 2, 3], "capabl": 2, "least": 2, "64gb": 2, "recommend": [2, 3], "howev": [2, 3], "depend": 2, "compil": 2, "yourself": 2, "develop": 2, "gnu": 2, "libquadmath": 2, "boost": 2, "sudoer": 2, "sudo": 2, "apt": 2, "dev": 2, "setuptool": 2, "pip": 2, "libboost": 2, "export": 2, "ld_library_path": 2, "yourpath": 2, "lib": 2, "Or": 2, "add": 2, "ld": 2, "so": 2, "conf": 2, "python": 2, "setup": 2, "py": 2, "As": [2, 3], "default": [0, 2, 3], "llvm": 2, "xcode": 2, "doe": [2, 3], "quadmath": 2, "time": 2, "being": 2, "we": [2, 3], "maco": 2, "either": [0, 3], "latter": 3, "unlock": 3, "potenti": 3, "invok": 3, "option": [0, 3], "displai": 3, "execut": 3, "h": 3, "note": 3, "script": 3, "possess": 3, "global": 3, "specif": 3, "specifi": [0, 3], "argument": 3, "choic": 3, "xscore": 3, "demo": 3, "eur": 3, "simul": 3, "out": 3, "yet": 3, "third": 3, "fourth": 3, "prepend": 3, "pw": 3, "gwasa": 3, "tsv": 3, "consult": 3, "point": 3, "chi2sum": [0, 1, 3], "particular": 3, "extend": 3, "50000": [0, 3], "varianc": [0, 3], "varcutoff": [0, 3], "99": [0, 3], "1000": 3, "project": 3, "convert": 3, "misc": 3, "folder": 3, "below": 3, "replac": 3, "38": 3, "37": 3, "get1kggrch38": 3, "sh": 3, "pathtostor": 3, "cpu": 3, "util": 3, "plink": 3, "1kg": 3, "want": 3, "retain": 3, "origin": 3, "linux": 3, "o": 3, "instanc": [0, 3], "load_refpanel": [0, 3], "been": 3, "gzip": [0, 3], "code": [0, 3], "do": [0, 3], "job": 3, "recod": 3, "12": [0, 3], "transpos": 3, "By": 3, "increas": 3, "abov": [0, 3], "subset": 3, "around": 3, "50gb": 3, "take": [0, 3], "hour": [0, 3], "therefor": 3, "your_filenam": 3, "lncrna": 3, "valid": 3, "match": [0, 3], "overlap": [0, 3], "both": 3, "dataset": 3, "need": [0, 3], "also": [0, 3], "possibl": 3, "document": 3, "instead": 3, "direct": [0, 3], "rscol": [0, 3], "respect": 3, "refernc": 3, "visual": 3, "plot_genesnp": [0, 3], "aoah": 3, "mark_window": [0, 3], "show_correl": [0, 3], "r": 3, "score_al": [0, 3], "21": [0, 3], "score_chr": [0, 3], "wdr12": 3, "farp2": 3, "section": 3, "read": 3, "r_success": 3, "r_fail": 3, "r_totalfail": 3, "nsnp": 3, "info": 3, "reason": 3, "successfulli": 3, "fail": [0, 3], "due": 3, "non": [0, 3], "converg": 3, "other": 3, "like": 3, "exact": 3, "most": 3, "suitabl": 3, "autorescor": [0, 3], "max": [0, 3], "enough": 3, "iter": 3, "eventu": 3, "ratio": [0, 3], "largest": 3, "smallest": 3, "eigenvalu": 3, "larg": 3, "mai": [0, 3], "veri": 3, "slowli": 3, "case": 3, "often": 3, "help": 3, "reduc": 3, "manual": 3, "rescor": [0, 3], "save_scor": [0, 3], "save": [0, 3], "pair": [0, 3], "manhattan": [0, 3], "plot": [0, 3], "plot_manhattan": [0, 3], "matplotlib": 3, "modifi": 3, "current": [0, 3], "pyplot": 3, "fulli": 3, "introduc": 3, "defin": 3, "load_scor": [0, 3], "initialiaz": 3, "pscorer": 3, "uniform": 3, "variabl": 3, "mont": 3, "carlo": 3, "directli": 3, "compar": 3, "randomli": 3, "cost": 3, "solei": 3, "m": [0, 3], "assum": 3, "member": 3, "subsequ": 3, "remov": [0, 3], "befor": 3, "These": 3, "mark": [0, 3], "nan": 3, "preprint": 3, "1101": 3, "2021": 3, "05": [0, 3], "16": [0, 3], "21257289": 3, "featur": 3, "peer": 3, "review": 3, "zsum": [0, 1, 3], "lefttail": [0, 3], "side": 3, "anti": 3, "standard": 3, "filenamea": 3, "filenameb": 3, "b": 3, "routin": 3, "necessari": 3, "beta": [0, 3], "matchallel": [0, 3], "jointli": [0, 3], "qq": [0, 3], "jointlyrank": [0, 3], "e_a": [0, 3], "e_b": [0, 3], "trim26": 3, "rsum": [0, 1, 3], "ge": 3, "yourannotationfil": 3, "gtex": [0, 3], "get_gtex_expr": [0, 3], "yourgtexfilenam": 3, "sever": [0, 3], "step": 3, "load_expr": [0, 3], "tpm": 3, "plot_genexpr": [0, 3], "csmd1": 3, "cdh13": 3, "tzscore": [0, 3], "cbar_po": [0, 3], "01": [0, 3], "similar": 3, "fashion": 3, "proxim": 3, "count": 3, "degre": 3, "freedom": 3, "To": 3, "pathwaynam": 3, "pval": 3, "pvalg1": 3, "fusion": 3, "last": 3, "lists": 3, "eqtl": 3, "chromatin": 3, "complement": 3, "associ": [0, 3], "rsid": [0, 3], "thei": 3, "joint": [0, 3], "preced": 0, "static": 0, "naid": 0, "na": 0, "log10p": 0, "300": 0, "textfil": 0, "split": 0, "log10": 0, "rank_mapp": 0, "save_gwa": 0, "get_topscor": 0, "top": 0, "get_geneinfo": 0, "detail": 0, "scoringresult": 0, "region": 0, "siglin": 0, "logsigthreshold": 0, "labelsig": 0, "labellist": 0, "style": 0, "color": 0, "produc": 0, "_score": 0, "str": 0, "horizont": 0, "label": 0, "log": 0, "design": 0, "classic": 0, "clean": 0, "obtain": 0, "previou": 0, "unloadref": 0, "intlimit": 0, "memori": 0, "integr": 0, "term": 0, "maf": 0, "1000000": 0, "activatefail": 0, "helper": 0, "forc": 0, "activ": 0, "success": 0, "know": 0, "what": 0, "scor": 0, "score_gene_bulk_chr": 0, "bulk": 0, "locat": 0, "format": 0, "rsid1": 0, "rsid2": 0, "gwasid1": 0, "gwasid2": 0, "matrix": 0, "col": 0, "tickspac": 0, "limegreen": 0, "corrcmap": 0, "correl": 0, "space": 0, "tick": 0, "cmap": 0, "colormap": 0, "test_gene_assocdir": 0, "epsilon": 0, "08": 0, "wchi2sum": [0, 1], "jointlyrank_mapp": 0, "invert": 0, "idcol": 0, "mincutoff": 0, "share": 0, "matchrefpanel": 0, "indic": 0, "matchalleles_mapp": 0, "zscore": 0, "pcolor": 0, "ncolor": 0, "darkviolet": 0, "neg": 0, "pcorr": 0, "Not": 0, "factor": 0, "6": 0, "7": 0, "8": 0, "11": 0, "13": 0, "14": 0, "15": 0, "17": 0, "19": 0, "20": 0, "score_map": 0, "v8": 0, "gene1": 0, "gene2": 0, "coordin": 0}, "objects": {"PascalX.genescorer": [[0, 0, 1, "", "chi2sum"], [0, 0, 1, "", "genescorer"], [0, 0, 1, "", "wchi2sum"]], "PascalX.genescorer.chi2sum": [[0, 1, 1, "", "activateFails"], [0, 1, 1, "", "plot_genesnps"], [0, 1, 1, "", "rescore"], [0, 1, 1, "", "score"], [0, 1, 1, "", "score_gene_bulk_chr"], [0, 1, 1, "", "test_gene_assocdir"]], "PascalX.genescorer.genescorer": [[0, 1, 1, "", "clean"], [0, 1, 1, "", "get_geneinfo"], [0, 1, 1, "", "get_topscores"], [0, 1, 1, "", "load_GWAS"], [0, 1, 1, "", "load_genome"], [0, 1, 1, "", "load_mapping"], [0, 1, 1, "", "load_refpanel"], [0, 1, 1, "", "load_scores"], [0, 1, 1, "", "matchAlleles"], [0, 1, 1, "", "plot_Manhattan"], [0, 1, 1, "", "rank"], [0, 1, 1, "", "rank_mapper"], [0, 1, 1, "", "save_GWAS"], [0, 1, 1, "", "save_scores"], [0, 1, 1, "", "score_all"], [0, 1, 1, "", "score_chr"]], "PascalX.genescorer.wchi2sum": [[0, 1, 1, "", "score"]], "PascalX.genexpr": [[0, 0, 1, "", "genexpr"]], "PascalX.genexpr.genexpr": [[0, 1, 1, "", "chi2rank"], [0, 1, 1, "", "get_GTEX_expr"], [0, 1, 1, "", "load_expr"], [0, 1, 1, "", "load_genome"], [0, 1, 1, "", "plot_genexpr"]], "PascalX.genome": [[0, 0, 1, "", "genome"]], "PascalX.genome.genome": [[0, 1, 1, "", "gene_info"], [0, 1, 1, "", "get_ensembl_annotation"], [0, 1, 1, "", "load_genome"]], "PascalX.mapper": [[0, 0, 1, "", "mapper"]], "PascalX.mapper.mapper": [[0, 1, 1, "", "load_mapping"]], "PascalX.pathway": [[0, 0, 1, "", "chi2perm"], [0, 0, 1, "", "chi2rank"]], "PascalX.pathway.chi2perm": [[0, 1, 1, "", "get_sigpathways"], [0, 1, 1, "", "load_modules"], [0, 1, 1, "", "score"], [0, 1, 1, "", "set_genescorer"]], "PascalX.pathway.chi2rank": [[0, 1, 1, "", "get_sigpathways"], [0, 1, 1, "", "load_modules"], [0, 1, 1, "", "score"], [0, 1, 1, "", "set_genescorer"]], "PascalX.refpanel": [[0, 0, 1, "", "refpanel"]], "PascalX.refpanel.refpanel": [[0, 1, 1, "", "getChrSNPs"], [0, 1, 1, "", "getSNPtoChrMap"], [0, 1, 1, "", "load_pos_reference"], [0, 1, 1, "", "load_snp_reference"], [0, 1, 1, "", "set_refpanel"]], "PascalX.snpdb": [[0, 0, 1, "", "db"]], "PascalX.snpdb.db": [[0, 1, 1, "", "close"], [0, 1, 1, "", "get"], [0, 1, 1, "", "getKeys"], [0, 1, 1, "", "getPosatSNPs"], [0, 1, 1, "", "getSNPKeys"], [0, 1, 1, "", "getSNPatPos"], [0, 1, 1, "", "getSNPs"], [0, 1, 1, "", "getSortedKeys"], [0, 1, 1, "", "insert"], [0, 1, 1, "", "open"]], "PascalX.xscorer": [[0, 0, 1, "", "rsum"], [0, 0, 1, "", "zsum"]], "PascalX.xscorer.rsum": [[0, 1, 1, "", "get_topscores"], [0, 1, 1, "", "jointlyRank"], [0, 1, 1, "", "jointlyRank_mapper"], [0, 1, 1, "", "load_GWAS"], [0, 1, 1, "", "load_genome"], [0, 1, 1, "", "load_mapping"], [0, 1, 1, "", "load_refpanel"], [0, 1, 1, "", "load_scores"], [0, 1, 1, "", "matchAlleles"], [0, 1, 1, "", "matchAlleles_mapper"], [0, 1, 1, "", "plot_genesnps"], [0, 1, 1, "", "save_scores"], [0, 1, 1, "", "score"], [0, 1, 1, "", "score_all"], [0, 1, 1, "", "score_chr"], [0, 1, 1, "", "score_map"]], "PascalX.xscorer.zsum": [[0, 1, 1, "", "get_topscores"], [0, 1, 1, "", "jointlyRank"], [0, 1, 1, "", "jointlyRank_mapper"], [0, 1, 1, "", "load_GWAS"], [0, 1, 1, "", "load_genome"], [0, 1, 1, "", "load_mapping"], [0, 1, 1, "", "load_refpanel"], [0, 1, 1, "", "load_scores"], [0, 1, 1, "", "matchAlleles"], [0, 1, 1, "", "matchAlleles_mapper"], [0, 1, 1, "", "plot_genesnps"], [0, 1, 1, "", "save_scores"], [0, 1, 1, "", "score"], [0, 1, 1, "", "score_all"], [0, 1, 1, "", "score_chr"], [0, 1, 1, "", "score_map"]]}, "objtypes": {"0": "py:class", "1": "py:method"}, "objnames": {"0": ["py", "class", "Python class"], "1": ["py", "method", "Python method"]}, "titleterms": {"api": 0, "document": [0, 1], "genom": 0, "genescor": 0, "pathwayscor": 0, "xscorer": 0, "genexprscor": 0, "snp": [0, 3], "databas": 0, "refer": 0, "panel": 0, "mapper": 0, "welcom": 1, "pascalx": 1, "": 1, "content": 1, "index": 1, "licens": 1, "citat": 1, "polici": 1, "contact": 1, "instal": 2, "multi": 2, "platform": 2, "docker": 2, "debian": 2, "ubuntu": 2, "mac": 2, "window": 2, "usag": 3, "command": 3, "line": 3, "interfac": 3, "gene": 3, "score": 3, "pathwai": 3, "x": 3, "tissu": 3}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 57}, "alltitles": {"Welcome to PascalX\u2019s documentation!": [[1, "welcome-to-pascalx-s-documentation"]], "Contents": [[1, null]], "Index": [[1, "index"]], "License and citation policy": [[1, "license-and-citation-policy"]], "Contact": [[1, "contact"]], "Installation": [[2, "installation"]], "Multi-platform (Docker)": [[2, "multi-platform-docker"]], "Debian/Ubuntu": [[2, "debian-ubuntu"]], "Mac": [[2, "mac"]], "Windows": [[2, "windows"]], "Usage": [[3, "usage"]], "Command-line interface": [[3, "command-line-interface"]], "Gene scoring": [[3, "gene-scoring"]], "Pathway scoring": [[3, "pathway-scoring"]], "X scoring": [[3, "x-scoring"]], "Tissue scoring": [[3, "tissue-scoring"]], "SNP-gene scoring": [[3, "snp-gene-scoring"]], "API documentation": [[0, "api-documentation"]], "Genome": [[0, "genome"]], "Genescorer": [[0, "genescorer"]], "Pathwayscorer": [[0, "pathwayscorer"]], "Xscorer": [[0, "xscorer"]], "Genexprscorer": [[0, "genexprscorer"]], "SNP database": [[0, "snp-database"]], "Reference panel": [[0, "reference-panel"]], "Mapper": [[0, "mapper"]]}, "indexentries": {"activatefails() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.activateFails"]], "chi2perm (class in pascalx.pathway)": [[0, "PascalX.pathway.chi2perm"]], "chi2rank (class in pascalx.pathway)": [[0, "PascalX.pathway.chi2rank"]], "chi2rank() (pascalx.genexpr.genexpr method)": [[0, "PascalX.genexpr.genexpr.chi2rank"]], "chi2sum (class in pascalx.genescorer)": [[0, "PascalX.genescorer.chi2sum"]], "clean() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.clean"]], "close() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.close"]], "db (class in pascalx.snpdb)": [[0, "PascalX.snpdb.db"]], "gene_info() (pascalx.genome.genome method)": [[0, "PascalX.genome.genome.gene_info"]], "genescorer (class in pascalx.genescorer)": [[0, "PascalX.genescorer.genescorer"]], "genexpr (class in pascalx.genexpr)": [[0, "PascalX.genexpr.genexpr"]], "genome (class in pascalx.genome)": [[0, "PascalX.genome.genome"]], "get() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.get"]], "getchrsnps() (pascalx.refpanel.refpanel method)": [[0, "PascalX.refpanel.refpanel.getChrSNPs"]], "getkeys() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getKeys"]], "getposatsnps() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getPosatSNPs"]], "getsnpkeys() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getSNPKeys"]], "getsnpatpos() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getSNPatPos"]], "getsnps() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getSNPs"]], "getsnptochrmap() (pascalx.refpanel.refpanel method)": [[0, "PascalX.refpanel.refpanel.getSNPtoChrMap"]], "getsortedkeys() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.getSortedKeys"]], "get_gtex_expr() (pascalx.genexpr.genexpr method)": [[0, "PascalX.genexpr.genexpr.get_GTEX_expr"]], "get_ensembl_annotation() (pascalx.genome.genome method)": [[0, "PascalX.genome.genome.get_ensembl_annotation"]], "get_geneinfo() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.get_geneinfo"]], "get_sigpathways() (pascalx.pathway.chi2perm method)": [[0, "PascalX.pathway.chi2perm.get_sigpathways"]], "get_sigpathways() (pascalx.pathway.chi2rank method)": [[0, "PascalX.pathway.chi2rank.get_sigpathways"]], "get_topscores() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.get_topscores"]], "get_topscores() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.get_topscores"]], "get_topscores() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.get_topscores"]], "insert() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.insert"]], "jointlyrank() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.jointlyRank"]], "jointlyrank() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.jointlyRank"]], "jointlyrank_mapper() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.jointlyRank_mapper"]], "jointlyrank_mapper() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.jointlyRank_mapper"]], "load_gwas() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.load_GWAS"]], "load_gwas() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.load_GWAS"]], "load_gwas() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.load_GWAS"]], "load_expr() (pascalx.genexpr.genexpr method)": [[0, "PascalX.genexpr.genexpr.load_expr"]], "load_genome() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.load_genome"]], "load_genome() (pascalx.genexpr.genexpr method)": [[0, "PascalX.genexpr.genexpr.load_genome"]], "load_genome() (pascalx.genome.genome method)": [[0, "PascalX.genome.genome.load_genome"]], "load_genome() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.load_genome"]], "load_genome() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.load_genome"]], "load_mapping() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.load_mapping"]], "load_mapping() (pascalx.mapper.mapper method)": [[0, "PascalX.mapper.mapper.load_mapping"]], "load_mapping() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.load_mapping"]], "load_mapping() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.load_mapping"]], "load_modules() (pascalx.pathway.chi2perm method)": [[0, "PascalX.pathway.chi2perm.load_modules"]], "load_modules() (pascalx.pathway.chi2rank method)": [[0, "PascalX.pathway.chi2rank.load_modules"]], "load_pos_reference() (pascalx.refpanel.refpanel method)": [[0, "PascalX.refpanel.refpanel.load_pos_reference"]], "load_refpanel() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.load_refpanel"]], "load_refpanel() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.load_refpanel"]], "load_refpanel() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.load_refpanel"]], "load_scores() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.load_scores"]], "load_scores() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.load_scores"]], "load_scores() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.load_scores"]], "load_snp_reference() (pascalx.refpanel.refpanel method)": [[0, "PascalX.refpanel.refpanel.load_snp_reference"]], "mapper (class in pascalx.mapper)": [[0, "PascalX.mapper.mapper"]], "matchalleles() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.matchAlleles"]], "matchalleles() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.matchAlleles"]], "matchalleles() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.matchAlleles"]], "matchalleles_mapper() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.matchAlleles_mapper"]], "matchalleles_mapper() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.matchAlleles_mapper"]], "open() (pascalx.snpdb.db method)": [[0, "PascalX.snpdb.db.open"]], "plot_manhattan() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.plot_Manhattan"]], "plot_genesnps() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.plot_genesnps"]], "plot_genesnps() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.plot_genesnps"]], "plot_genesnps() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.plot_genesnps"]], "plot_genexpr() (pascalx.genexpr.genexpr method)": [[0, "PascalX.genexpr.genexpr.plot_genexpr"]], "rank() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.rank"]], "rank_mapper() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.rank_mapper"]], "refpanel (class in pascalx.refpanel)": [[0, "PascalX.refpanel.refpanel"]], "rescore() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.rescore"]], "rsum (class in pascalx.xscorer)": [[0, "PascalX.xscorer.rsum"]], "save_gwas() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.save_GWAS"]], "save_scores() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.save_scores"]], "save_scores() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.save_scores"]], "save_scores() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.save_scores"]], "score() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.score"]], "score() (pascalx.genescorer.wchi2sum method)": [[0, "PascalX.genescorer.wchi2sum.score"]], "score() (pascalx.pathway.chi2perm method)": [[0, "PascalX.pathway.chi2perm.score"]], "score() (pascalx.pathway.chi2rank method)": [[0, "PascalX.pathway.chi2rank.score"]], "score() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.score"]], "score() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.score"]], "score_all() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.score_all"]], "score_all() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.score_all"]], "score_all() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.score_all"]], "score_chr() (pascalx.genescorer.genescorer method)": [[0, "PascalX.genescorer.genescorer.score_chr"]], "score_chr() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.score_chr"]], "score_chr() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.score_chr"]], "score_gene_bulk_chr() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.score_gene_bulk_chr"]], "score_map() (pascalx.xscorer.rsum method)": [[0, "PascalX.xscorer.rsum.score_map"]], "score_map() (pascalx.xscorer.zsum method)": [[0, "PascalX.xscorer.zsum.score_map"]], "set_genescorer() (pascalx.pathway.chi2perm method)": [[0, "PascalX.pathway.chi2perm.set_genescorer"]], "set_genescorer() (pascalx.pathway.chi2rank method)": [[0, "PascalX.pathway.chi2rank.set_genescorer"]], "set_refpanel() (pascalx.refpanel.refpanel method)": [[0, "PascalX.refpanel.refpanel.set_refpanel"]], "test_gene_assocdir() (pascalx.genescorer.chi2sum method)": [[0, "PascalX.genescorer.chi2sum.test_gene_assocdir"]], "wchi2sum (class in pascalx.genescorer)": [[0, "PascalX.genescorer.wchi2sum"]], "zsum (class in pascalx.xscorer)": [[0, "PascalX.xscorer.zsum"]]}})